Code/GraphMol/Descriptors/USRDescriptor.cpp

```diff
@@ -0,0 +1,326 @@
+//       copyright notice, this list of conditions and the following
+//       disclaimer in the documentation and/or other materials provided
+//       with the distribution.
+//     * Neither the name of Novartis Institutes for BioMedical Research Inc.
+//       nor the names of its contributors may be used to endorse or promote
+//       products derived from this software without specific prior written
+//       permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include <Geometry/point.h>
+#include <Numerics/Vector.h>
+#include <GraphMol/RDKitBase.h>
+#include <GraphMol/SmilesParse/SmilesParse.h>
+#include <GraphMol/Substruct/SubstructMatch.h>
+#include <boost/foreach.hpp>
+#include "USRDescriptor.h"
+
+#include <boost/flyweight.hpp>
+#include <boost/flyweight/key_value.hpp>
+#include <boost/flyweight/no_tracking.hpp>
+
+namespace RDKit {
+
+namespace {
+void calcDistances(const RDGeom::Point3DConstPtrVect &coords,
+                   const RDGeom::Point3D &point,
+                   std::vector<double> &distances) {
+  distances.resize(coords.size());
+  unsigned int i = 0;
+  // loop over coordinates
+  BOOST_FOREACH (const RDGeom::Point3D *tpp, coords) {
+    distances[i++] = (*tpp - point).length();
+  }
+}
+
+void calcCentroid(const RDGeom::Point3DConstPtrVect &coords,
+                  RDGeom::Point3D &pt) {
+  PRECONDITION(!coords.empty(), "no coordinates");
+  // set pt to zero
+  pt *= 0.0;
+  // loop over coordinates
+  BOOST_FOREACH (const RDGeom::Point3D *opt, coords) { pt += *opt; }
+  pt /= coords.size();
+}
+
+unsigned int largestValId(const std::vector<double> &v) {
+  PRECONDITION(!v.empty(), "no values");
+  double res = v[0];
+  unsigned int id = 0;
+  for (unsigned int i = 1; i < v.size(); ++i) {
+    if (v[i] > res) {
+      res = v[i];
+      id = i;
+    }
+  }
+  return id;
+}
+
+unsigned int smallestValId(const std::vector<double> &v) {
+  PRECONDITION(!v.empty(), "no values");
+  double res = v[0];
+  unsigned int id = 0;
+  for (unsigned int i = 1; i < v.size(); ++i) {
+    if (v[i] < res) {
+      res = v[i];
+      id = i;
+    }
+  }
+  return id;
+}
+
+void calcMoments(const std::vector<double> &dist,
```

Comment:
Perhaps the moments calculations should be in a central location as it is potentially used outside of USR?